### ISSUE TITLE
Timeline daily hour interval

### ISSUE THREAD

**Issue by stochris** [namespace-ee/react-calendar-timeline#685]
**Is your feature request related to a problem? Please describe.**
On a project I'm working the end user requests hour intervals for the timeline
**Describe the solution you'd like**
I'd like to have prop ( or set of props ) which can be used to set the hour interval

**Additional context**

This is a feature I'd be more than open to implementing. However, I would need support / advice from someone on the team, regarding where and how.

Ideally it would look like this:

```
<Timeline ... hourInterval={{ start: 8, end: 18 }} />
```

Would that be feasible ?

**Comment by Ilaiwi** [comment]
@stochris I am not quite sure I get the feature you want to implement from the description.

have you tried using visible time start and visible time end? also for the headers you can use Timeline headers and give DateHeader a unit hour to only display hours on the header

**Comment by stochris** [comment]
Sorry, what I meant is limiting the hour interval for each day to, for example, working hours. So, when you zoom in, it doesn't display all hour intervals, but only from 9 to 18.

I looked a bit around the code, and noticed there are quite a few things that would need changing.

**Comment by Ilaiwi** [comment]
@stochris I don't think this feature is possible. The timeline is built on the assumption that a day is 24 hours

**Comment by stochris** [comment]
Thanks, I'm gonna have to try anyway 😅

**Comment by otang** [comment]
@stochris I need the same thing. Did you find any solution yet?

**Comment by Kevmue** [comment]
This would be a great feature to make the timeline more compact!

**Comment by 2f-concept** [comment]
I'd like to have this feature too ! If someone find the solution, please share ! ;)

**Comment by Ilaiwi** [comment]
@otang @Kevmue  @2f-concept lmk know guys if you need any help implementing this. This is still not a priority for this library.

**Comment by Ruzen1** [comment]
I want this too....

**Comment by magarbikram** [comment]
@Ilaiwi I am also looking for a way to have this feature, I am willing to contribute if I get some guideline on this project

**Comment by Ilaiwi** [comment]
@magarbikram sure thing.

**Comment by magarbikram** [comment]
@Ilaiwi  thanks, how can we proceed?, my email address is magar.bik@gmail.com. Lets connect there

**Comment by magarbikram** [comment]
So I have started working on it and based on my initial evaluation, it doesn't seems possible, any guidance would be great. 

- I am able to remove the closed hours column
- But unable to distribute remaining space to open hours column

Any idea from you @Ilaiwi  @stochris

**Comment by Ilaiwi** [comment]
@magarbikram will make sure to get back to this on Monday since I am on Holiday right now. Thanks for investigating

**Comment by Ilaiwi** [comment]
@magarbikram if you scroll right will other columns (days) render or it's all empty?
The main issue here we are relaying on days being 24hours and all the calculations are made with this assumption hence all the `startOf(`day`)` `endOf(`day`)` all over the project. The only way I am thinking this might be possible is to first do a refactor to abstract `moment` calculations in a util module something like what `MuiPickersUtilsProvider` do in [material-ui-picker](https://material-ui-pickers.dev/guides/formats). Then we can replace the helpers with your custom calculations to provide new `start` and `end` of day

**Comment by magarbikram** [comment]
It will render on the right as well, but there is a gap caused by closed hours. The main issue I am having is not having a way to fill those gaps (may be with increased width per hour), but since we map each moment with exact position (positionX), it's hard to imagine how to do it technically/logically

**Comment by magarbikram** [comment]
@Ilaiwi I like the idea of abstracting the calculation in separate provider, but I am still not clear on how can we distribute the space from a gap by closed hours to open hours. For eg: in 3day view
-------------day 1 ------------------------ day 2------------------------  day 3----------------
7am--------------------- 7pm--7am-----------------7pm-- 7am-------------------------7pm

the gap between 7pm - 7am

with this function
```
export function calculateXPositionForTime(
  canvasTimeStart,
  canvasTimeEnd,
  canvasWidth,
  time
) {

}
```

**Comment by Ilaiwi** [comment]
@magarbikram not sure I follow

**Comment by magarbikram** [comment]
@Ilaiwi, what I was saying is all the calculation is based on calculateXPositionForTime function which linearly distribute time in a canvas width, the problem I am having is how to distribute the gap that would be created by the closed hours. 

In iterateTimes function, I added a logic to exclude columns which are not open, and this creates a gap in timeline
```
export function iterateTimes(start, end, unit, timeSteps, callback) {
  let time = moment(start).startOf(unit)

  if (timeSteps[unit] && timeSteps[unit] > 1) {
    let value = time.get(unit)
    time.set(unit, value - value % timeSteps[unit])
  }
  const dayTimeStart = 7;
  const dayTimeEnd = 19;
  while (time.valueOf() < end) {
    let nextTime = moment(time).add(timeSteps[unit] || 1, `${unit}s`)
    if(nextTime.hour()>dayTimeStart && nextTime.hour()<=dayTimeEnd){
     callback(time, nextTime)
    }
    time = nextTime
  }
}
```
![image](https://user-images.githubusercontent.com/493559/120685554-cec13e80-c454-11eb-8474-c5312d9bcc9b.png)
